### DISCUSSION TITLE
Prophet - how to account for this daily seasonality?

### DISCUSSION THREAD

**Discussion by jjhuang4** [alan-turing-institute/sktime#7681] [Q&A]
I'm trying to experiment with the Prophet model on a electricity demand dataset [(link here)](https://www.kaggle.com/datasets/vstacknocopyright/electricity)

This is a dataset containing 45,000+ observations recorded every half an hour,  so every increment of 48 represents electricity demand updated every half hour over a 24 hour period. I've found one pattern to repeat throughout the time series, with slight shifts explained by changes to the trend from seasonal weather. I think Prophet accounts for the yearly/monthly seasonality from season changes well enough by default.

My question: How do I go about encoding this hourly/daily seasonality (not sure what to call it) into my model? (pictured is two periods of said pattern). I'm so far just forecasting based on the time series alone (without any covariates/regressors).

![image](https://github.com/user-attachments/assets/e8c79f4e-520c-487c-9e9e-2c6dcec968ae)

I've tried looking at some documentation and examples and so far have found some promise in setting daily_seasonality=False and including add_seasonality() in my model arguments as follows:

```txt
forecaster = Prophet(
    seasonality_mode='additive',
    add_country_holidays={'country_name': 'Australia'},
    yearly_seasonality=True,
    weekly_seasonality=True,
    daily_seasonality=False,
    add_seasonality={
        'name': 'daily-granularity',
        'period': 1, # period of 1 for daily seasonality
        'fourier_order': 4
    }
)
```

However, I'm lost as to how I should decide which fourier_order parameter to use, and how to verify whether the seasonality I added would capture the exact pattern I see in my time series without "brute force" testing the forecasted values against a test set.

I'm pretty novice to this discipline, so any source or advice is appreciated!

**Comment by felipeangelimvieira** [comment]
Hi jjhuang! First, how familiar you are to how Prophet models the seasonality? The `fourier_order` is basically the number of pairs of sines and cosines with the specified period, that are internally added by Prophet as exogenous regressors. In your case, I believe the period should be the number of observations on each day (48). 

To determine the `fourier_order`, you can tune it on the test set using one of the tuners that sktime offers, or analyze the in-sample residues (y_pred - y_true), and check if these residues have a seasonal pattern. If yes, it means that there is seasonal information that is not being captured by your model, and you should probably consider increasing the number of fourier terms, or verify if there's something else, such as seasonality that changes over time.

**Reply by jjhuang4** [reply]
Hi Felipe, thanks for getting back to me so quickly.

I've tried your suggestion and changed my period to 48, and fourier_order to 12, and immediately saw large improvements. My forecasted values now capture the 24 hour cycle pattern when it didn't before. I'll continue to fine tune the fourier_order to see if I can get better results. Here's an updated plot of several periods:

![image](https://github.com/user-attachments/assets/e76d17ca-38fa-47cb-835c-e20982accb35)


My understanding is that the period is the number of observations that it takes for one seasonal cycle to complete, so using a period of 48 makes sense for making daily seasonality. I now want to account for the slight shifts in these "daily" cycles caused by variation in days of the week / seasonal changes. I have weekly_seasonality=True and yearly_seasonality=True in my current forecaster, but I'm seeing residuals (y_test-y_pred) that look like this:

![image](https://github.com/user-attachments/assets/ba4ae968-b0db-4ac7-9b39-541cd076c2bf)
(Showing 48*7 = 336 observations representing a whole week, so one period of the "weekly" seasonality).
So I'm assuming that this residual seasonality isn't being captured/fitted by weekly_seasonality in the model as it is currently.

A couple of new questions:

What "period" is the default for yearly_seasonality, weekly_seasonality, daily_seasonality, etc? From https://github.com/facebook/prophet/issues/1747 it looks like there are default fourier terms but not so much a default period; I'm just curious because it looks like the way my data is structured is messing these up, making it so I have to add custom seasonalities to account for different periods (i.e, for most time series the lowest level of granularity is at the daily level, but mine has at the half-hour level).

And also, how do I add multiple new seasonalities? I want to add a new custom seasonality with period 336, but can't find documentation for doing so (most of what I can find is using the base fbProphet, but the sktime wrapper seems to handle add_seasonality a little differently). My current code is:

```txt
y = electricity['nswdemand']
y_train, y_test = temporal_train_test_split(y, test_size=0.2)
forecaster = Prophet(
    seasonality_mode='additive',
    add_country_holidays={'country_name': 'Australia'},
    yearly_seasonality=True,
    weekly_seasonality=True,
    daily_seasonality=True,
    add_seasonality={
        'name': 'daily-cycle',
        'period': 48, # period of 48 representing observations per day
        'fourier_order': 12
    },
)
forecaster.fit(y_train)
```
(There actually does seem to be support for adding multiple custom seasonalities as of [this PR ](https://github.com/sktime/sktime/pull/1378) but I don't see it reflected in the sktime Prophet documentation).

**Reply by jjhuang4** [reply]
Update: I've found how to add multiple custom seasonalities by looking at the commit changes in #1378; it's just adding a list of dictionaries.
